morituri/common/simplefreedb.py

```diff
@@ -0,0 +1,137 @@
+        self._hello = urllib.quote_plus(hello)
+        self._slash_split_regex = re.compile(r'(?<!\\)/')
+
+    def _cddb_cmd(self, cmd):
+        cmd_arg = urllib.quote_plus('cddb ' + cmd)
+        url = '%s?cmd=%s&hello=%s&proto=%d' % (self.URL, cmd_arg, self._hello, self.PROTO)
+        req = urllib.urlopen(url)
+        return req.read().decode('utf-8')
+
+    @staticmethod
+    def _get_code(line):
+        return int(line.split(None, 1)[0])
+
+    def _split_dtitle(self, dtitle):
+        # Note: we can not use a simple dtitle.split('/') here because the
+        # slash could be escaped.
+        artist, title = re.split(self._slash_split_regex, dtitle, maxsplit=1)
+        return artist.rstrip(), title.lstrip()
+
+    def _craft_match(self, categ, discid_str, dtitle):
+        artist, title = self._split_dtitle(dtitle)
+        return {'category': categ,
+                'discid': int(discid_str, 16),
+                'artist_title': dtitle,
+                'artist': artist,
+                'title': title}
+
+    def query(self, discid, ntrks, offsets, nsecs):
+        cmd = 'query %08x %d %s %d' % (discid, ntrks, ' '.join(str(x) for x in offsets), nsecs)
+        data = self._cddb_cmd(cmd)
+        lines = data.splitlines()
+        code = self._get_code(lines[0])
+        matches = []
+        if code == 200:
+            line = lines[0]
+            matches.append(self._craft_match(*line.split(None, 3)[1:]))
+        elif code in (210, 211):
+            for line in lines[1:]:
+                if line == '.':
+                    break
+                matches.append(self._craft_match(*line.split(None, 2)))
+        return matches
+
+    def read(self, categ, discid):
+        cmd = 'read %s %08x' % (categ, discid)
+        data = self._cddb_cmd(cmd)
+        lines = data.splitlines()
+        code = self._get_code(lines[0])
+        if code != 210:
+            return None
+
+        data = {}
+        trackn = 0
+
+        for line in lines[1:]:
+            if line == '.':
+                break
+            if not line or line.startswith('#'):
+                continue
+            key, value = line.split('=', 1)
+            if not value:
+                continue
+            if key == 'DTITLE':
+                artist, title = self._split_dtitle(value)
+                data['artist_title'] = value
+                data['artist'] = artist
+                data['title'] = title
+            elif key == 'DYEAR':
+                data['year'] = int(value)
+            elif key == 'DGENRE':
+                data['genre'] = value
+            elif key.startswith('TTITLE'):
+                n = int(key[6:])
+                if n != trackn:
+                    raise Exception('Tracks titles are not monotically incrementing')
+                trackn += 1
+                if not n:
+                    data['tracks'] = []
+                data['tracks'].append(value)
+
+        return data
+
+def main():
+    test_queries = (
```

Comment:
I think it would be preferable to decouple testing from the actual code by creating an accompanying test module, like the rest of the application.

Reply:
This is not actually a test but more a debug tool (it's not testing output, it's dumping data). I'd prefer to keep tests depending on external websites such as freedb.org outside of the test suite.